drivers/adc/adc_sam0.c

```diff
@@ -328,6 +317,18 @@ static int start_read(const struct device *dev,
 	}
 
 	adc->AVGCTRL.reg = ADC_AVGCTRL_SAMPLENUM(sequence->oversampling);
+	switch (sequence->oversampling) {
```

Comment:
You shoud create an enum to store the `oversampling` value and do not add a default on the switch. Compiler will ensure that all valid entries defined on the enum will be on the switch case.

Reply:
The oversampling value is a numeric value from the DT which could ofc. be converted to an enum but I feel it would be an overkill to cast it just for the switch statement. 

The default is modeled after how the peripheral works. All oversampling values above 4 have a constant adjustment value. An argument could be made that maybe  a 
```c
if (oversampling < 4) {} else {} 
```
would be more suited here.

Reply:
Hi @guenzel-kinexon ,

Let's forget for a moment the implementation details and focus on the fix.

I was looking on the datasheet and it is not clear to me what you want to achieve. Could you explain in details what you want to achieve?

I see main sections (below) to explain how to use those bits. They should be combined with AVGCTRL.SAMPLENUM[3:0] but you are not changing then.
45.6.2.9 Accumulation
45.6.2.10 Averaging
45.6.2.11 Oversampling and Decimation

Reply:
Ultimately I want to use the peripheral to do hardware-averaging. To make this work the registers for accumulation as well as averaging have to be set.

The accumulation is already done in the driver (line 319, right above the switch statement). What is missing, is the averaging configuration, which I added in this PR.
As described in Accumulation section of the datasheet (chapter 33.6.6 for the SAMD21 datasheet), up to 16 samples the hardware does no automatic division. For 32 samples it already adds a factor but that is expected that the user already set the correct factor in the ADJRES field (as described in the Averaging chapter).

The datasheet is not super clear about this imho. but with trial&error on real hardware I figured it out to be a manual+automatic scheme requiring the driver to set the adjustment factor from 0..4 for averaging up to 16 samples and then let the hardware handle additional division for averaging factors beyond that.

Reply:
Ok, I see now. Thank you for the fix.

Yes, you can use` if (oversampling < 4) {} else {} `